docs: adding a faq document in preparation for Consul Enterprise 1.10.0 (#10373)
hashicorp/consul: website/content/docs/enterprise/license/faq.mdx

This will no longer work since each server must be able to find a valid license during the startup process.

## Q: Is there a tutorial available for the license configuration steps?

Please visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise).

## Q: What resources are available?

The list below is a great starting point for learning more about the license changes introduced in Consul Enterprise v1.10.0+ent.

- [Consul Enterprise Upgrade Documentation](https://www.consul.io/docs/enterprise/upgrades)

- [Consul License Documentation](http://localhost:3000/docs/enterprise/license)

- [License configuration values documentation](http://localhost:3000/docs/enterprise/license/overview#binaries-without-built-in-licenses)

- [Install a HashiCorp Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise)

## Q: Do these changes impact all customers/licenses?

The license changes introduced in v1.10.0 only affect Consul Enterprise.

> This impacts customers that have an enterprise binarie (EVAL / non-EVAL licenses) downloaded from releases.hashicorp.com.

Reviewer: ```suggestion
This impacts customers that have an enterprise binary (evaluation or non-evaluation licenses) downloaded from <releases.hashicorp.com>.
```

The license changes do not impact customers with the baked-in licensed binaries. In a later release of Consul Enterprise, baked-in binaries will be deprecated.

## Q: What is the product behavior change introduced by the licensing changes?

Starting with Consul Enterprise v1.10.0+ent, a valid license is required on-disk (auto-loading) or as an environment variable for Consul Enterprise to successfully boot-up.
The in-storage license feature will not be supported starting with Consul Enteprise v1.10.0+ent. All Consul Enterprise clusters using v1.10.0+ent must ensure that there is a valid license on-disk (auto-loaded) or as an environment variable.

> ## Q: What is the impact on EVAL licenses due to this change?

Reviewer: ```suggestion
## Q: What is the impact on evaluation licenses due to this change?
```

> The 6-hour trial period for EVAL licenses will be deprecated as of Consul Enterprise 1.10.0.  

Reviewer: ```suggestion
The 6-hour trial period for evaluation licenses will be deprecated as of Consul Enterprise 1.10.0.  
```

This means that any clusters deployed with Consul v1.10.0+ent binaries will need to have a valid license on the disk (auto-loaded) or as an environment variable.

> Failure in providing a valid license key will result in the Consul server agent not starting.

Reviewer: ```suggestion
Failure to provide a valid license key will result in the Consul server agent not starting.
```

## Q: Is there a grace period when licenses expire?

A license includes an `expiration_date` and a `termination_date`. An enteprise binary will cease to function once the `termination_date` has passed.

> Licenses will now have a 24 hrs grace period. The grace period is the time between license expiry till it terminates.

Reviewer: ```suggestion
Licenses will now have a 24 hrs grace period. The grace period is the time between license expiry until the time it terminates.
```

As Consul Enterprise approaches the expiration date, warnings will be issued in the system logs.

## Q: Does this affect client agents?

For existing clusters, if the clients agents are using ACLs and have a valid token, then they will be able to retrieve the license from the server.
If the client agents are not using ACLs, then the client agents will be need to have the license on-disk (auto-loading) or as an environment variable.
For new Consul clusters using Consul v1.10.0+ent, customers must ensure that there is a valid license on-disk (auto-loaded) or as an environment variable.

## Q: Does this affect snapshot agents?

Same behavior as Consul clients. See answer for [Does this affect snapshot agents? ](faq#q-does-this-affect-client-agents)

> ## Q: What is the behavior is the license is missing?

Reviewer 2: ```suggestion
## Q: What is the behavior if the license is missing?
```

Consul server agents will detect the absence of the license and immediately exit.

> Consul client agents will attempt to retrieve the license from servers if certain conditions are met: ACLs are enabled, a ACL token is provided to the client agent, the client agents configuration contains `start_join/retry_join` addresses, the start/retry join addresses are addresses of the Consul servers.

Reviewer: ```suggestion
Consul client agents will attempt to retrieve the license from servers if certain conditions are met:

* ACLs are enabled.
* An ACL token is provided to the client agent.
* The client agents configuration contains `start_join/retry_join` addresses.
* The start/retry join addresses are addresses of the Consul servers.
```

> Consul snapshot agents will attempt to retrieve the license from servers if certain conditions are met: ACLs are enabled, a ACL token is provided to the client agent, the client agents configuration contains `start_join/retry_join` addresses, the start/retry join addresses are addresses of the Consul servers.

Reviewer: This is the same set of license retrieval conditions as listed for the Consul clients. Should these two sections be combined into one?

## Q: Where can users get a trial license for Consul Enterprise?

Visit https://www.consul.io/trial for a free 30-day trial license.

## Q: Are the license files locked to a specific cluster?

The license files are not locked to a specific cluster or cluster node. The above changes apply to all nodes in a cluster.

## Q: Will this impact HCP Consul?

This will not impact HCP Consul.

## Q: Does this need to happen every time a node restarts, or is this a one-time check?

> Consul Enterprise binaries starting with v1.10.0+ent, will be subject to EULA check. Release v1.10.0+ent introduces the EULA check for eval licenses (non-EVAL licenses already go through EULA check during contractual agreement).

Reviewer: ```suggestion
Consul Enterprise binaries starting with v1.10.0+ent, will be subject to EULA check. Release v1.10.0+ent introduces the EULA check for evaluation licenses (non-evaluation licenses already go through EULA check during contractual agreement).
```

The agreement to a EULA happens only once (when the user gets their license), Consul Enterprise **will check for the presence of a valid license every time a node restarts**.

When a customer upgrades existing clusters to a v1.10.0+ent release, they need to have a valid license to successfully upgrade. This valid license must be auto-loaded.

When a customer deploys new clusters to a v1.10.0+ent release, they need to have a valid license to successfully upgrade. This valid license must be on-disk (auto-loaded).

## Q: What are the scenarios that a customer must plan for because of these changes?

New Consul cluster deployments using v1.10.0+ent will need to have a valid license to successfully deploy.
This valid license must be on-disk (auto-loaded) or as an environment variable.

> ## Q: What is the migration path for customers who want to migrate from their existing license-as-applied-via-the-CLI flow to the license on disk flow?

Reviewer 2: We should split this up between kube and vms. For kube you can just run your `helm upgrade` (cc @thisisnotashwin to double check).

Author: @lkysow @thisisnotashwin 

Would it be as simple as adding a K8s title header and providing this content

1. Run [`consul license get -signed`](/commands/license#get) to extract the license from their running cluster. Store the license in a secure location on disk.
1. Set up the necessary configuration so that when Consul Enterprise reboots it will have access to the required license. This could be via the client agent configuration file or an environment variable.
1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Proceed the `helm` [upgrade instructions](/docs/k8s/upgrade)

Reviewer 2: For k8s they should just need to upgrade the Helm chart to 0.32.0 when they upgrade to Consul 1.10.0. They don't need to extract the license at all.

> Run consul `license get -signed` to extract the license from their running cluster. Store the license in a secure location on disk.

Reviewer: ```suggestion
1. Run [`consul license get -signed`](https://www.consul.io/commands/license#get) to extract the license from their running cluster. Store the license in a secure location on disk.
```

> Set up the necessary configuration so that when Consul Enterprise reboots it will have access to the required license. This could be via the client agent configuration file or an environment variable.
> Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
> Proceed with the upgrade as normal.

Reviewer: ```suggestion
1. Set up the necessary configuration so that when Consul Enterprise reboots it will have access to the required license. This could be via the client agent configuration file or an environment variable.
1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Proceed with the upgrade as normal.
```

## Q: What is the migration path for customers who want to migrate from their existing perpetually-licensed binaries to the license on disk flow?

Aquire a valid Consul Enterprise license. If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
Store the license in a secure location on disk.
Set up the necessary configuration so that when Consul Enterprise reboots it will have the required license. This could be via the client agent configuration file or an environment variable.

> Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.

Reviewer: Are there any plans to introduce a vanity URL that contains "consul" in the path instead of "nomad"?

Author: Unfortunately @blake with how Learn is setup at the moment we don't have the ability to provide generic URLs, which is really what we need. We have requested this enhancement from the platform team.

> Aquire a valid Consul Enterprise license. If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
> Store the license in a secure location on disk.
> Set up the necessary configuration so that when Consul Enterprise reboots it will have the required license. This could be via the client agent configuration file or an environment variable.
> Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
> Proceed with the upgrade as normal.

Reviewer: ```suggestion
1. Acquire a valid Consul Enterprise license. If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
1. Store the license in a secure location on disk.
1. Set up the necessary configuration so that when Consul Enterprise reboots it will have the required license. This could be via the client agent configuration file or an environment variable.
1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Proceed with the upgrade as normal.
```

1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Proceed with the upgrade as normal.

### Kubernetes

1. Acquire a valid Consul Enterprise license. If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
1. Store the license in a secure location on disk.
1. Set up the necessary configuration so that when Consul Enterprise reboots it will have the required license. This could be via the client agent configuration file or an environment variable.
1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Proceed with the `helm` [upgrade instructions](/docs/k8s/upgrade)

> ## Q: Will Consul downgrades/rollbacks work?

Reviewer 3: We should call out here the bug we encountered on cloud-consul when trying to call the PutLicense call with a mixed version cluster.
 
Node A: v1.9.5.
Node B: v1.10.0

The call was made to v1.9.5 but since the leader was Node B when the call to PutLicense was redirected to Node B the PutLicense op failed. It's likely if customers downgrade and the leadership hasn't transferred back over to < 1.10.X that they will run into this issue.

Author: @chapmanc Thoughts on this ?

## Q: Are there potential pitfalls when downgrading or upgrading?

Assume a scenario where there are three Consul server nodes:
 Node A: v1.9.5 
 Node B: v1.10.0 [Leader]
 Node C: v1.9.5 

The command `consul license put` is issued from Node A. This will result in an error due to how Consul routes calls to the server node, Node B in this example.
Because Node A is a follower when the call to `consul license put` is issued, the call will be redirected to Node B (leader). 
The `consul license put` operation will fail due to being removed from v1.10.0. 
This is a scenario that could occur if a customer downgrades from 1.10.0+ and the Consul leadership has not transferred back over to a node not running 1.10.0+.
This also has the potential of occurring when upgrading if scheduled license updates or autoscaling groups recoveries are in place.

## Q: What are the scenarios that a customer must plan for because of these changes?

New Consul cluster deployments using 1.10.0+ent will need to have a valid license to successfully deploy.
This valid license must be on-disk (auto-loaded) or as an environment variable.

## Q: What is the migration path for customers who want to migrate from their existing license-as-applied-via-the-CLI flow to the license on disk flow?

### VM

1. Run [`consul license get -signed`](/commands/license#get) to extract the license from their running cluster. Store the license in a secure location on disk.
1. Set up the necessary configuration so that when Consul Enterprise reboots it will have access to the required license. This could be via the client agent configuration file or an environment variable.
1. Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
1. Follow the Consul upgrade [docummentation](/docs/upgrading).

### Kubernetes

> 1. Acquire a valid Consul Enterprise license. If you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license.
> 1. Set up the necessary configuration so that when Consul Enterprise reboots it will have access to the required license. This could be via the client agent configuration file or an environment variable.
>    Visit the [Enterprise License Tutorial](https://learn.hashicorp.com/tutorials/nomad/hashicorp-enterprise-license?in=consul/enterprise) for detailed steps on how to install the license key.
> 1. Proceed with the `helm` [upgrade instructions](/docs/k8s/upgrade).
> 

Reviewer 2: ```suggestion
**NOTE:** If you are not upgrading Consul or your Helm chart version then no action is required.

1. In order to use Consul Enterprise 1.10.0 or greater on Kubernetes you must use version 0.32.0 or greater of the Helm chart.
1. You should already have a Consul Enterprise license set as a Kubernetes secret. If you do not, then if you are an existing HashiCorp enterprise customer you may contact your organization's customer success manager (CSM) or email support-softwaredelivery@hashicorp.com for information on how to get your organization's enterprise license. Once you have the license then create a Kubernetes secret containing the license as described in [Kubernetes - Consul Enterprise](/docs/k8s/installation/deployment-configurations/consul-enterprise).
1. Follow the [Kubernetes Upgrade Docs](/docs/k8s/upgrade) to upgrade. No changes to your `values.yaml` file are needed to enable enterprise autoloading since this support is built in to consul-helm 0.32.0 and greater.

~> **WARNING:** If you are upgrading the Helm chart but **not** upgrading the Consul version, you must set `server.enterpriseLicense.enableLicenseAutoload: false`. See [Kubernetes - Consul Enterprise](/docs/k8s/installation/deployment-configurations/consul-enterprise) for more details.
```
